channeldb/kvdb/etcd/stm.go

```diff
@@ -235,18 +237,27 @@ func runSTM(s *stm, apply func(STM) error) error {
 		err     error
 	)
 
-	// In a loop try to apply and commit and roll back
-	// if the database has changed (CommitError).
+	// In a loop try to apply and commit and roll back if the database has
+	// changed (CommitError).
 	for {
-		// Abort STM if there was an application error.
+		select {
+		// Check if the STM is aborted and break the retry loop if it is.
+		case <-s.options.ctx.Done():
+			err = fmt.Errorf("aborted")
+			break
```

Comment:
We should add a test for this. I _think_ this'll actually break out of the _switch_ statement, and not the outer for loop unless a label is used.

Reply:
Alternatively, it can just return here.

Reply:
Thanks for pointing out, I wasn't aware of the label trick. Added a test for abort context.

Reply:
Yeah it's one of those subtle gotchas with switch statements and for loops. I've spent a lot of time in the past debugging issues where the root cause was the lack of a goroutine breaking out of a loop as expected.